File: 17_ORM & Code Structure/Task 01/controllers/packageController.js
const Package = require('../models/package');

// Get all packages
exports.getAllPackages = async (req, res) => {
  try {
    const packages = await Package.findAll();
    res.json(packages);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

// Get package by ID
exports.getPackageById = async (req, res) => {
  try {
    const package = await Package.findByPk(req.params.id);
    if (!package) {
      return res.status(404).json({ message: 'Package not found' });
    }
    res.json(package);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

// Add a new package
exports.addPackage = async (req, res) => {
  try {
    const newPackage = await Package.create(req.body);
    res.status(201).json(newPackage);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

// Update a package
exports.updatePackage = async (req, res) => {
  try {
    const package = await Package.findByPk(req.params.id);
    if (!package) {
      return res.status(404).json({ message: 'Package not found' });
    }
    await package.update(req.body);
    res.json(package);
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

// Delete a package
exports.deletePackage = async (req, res) => {
  try {
    const package = await Package.findByPk(req.params.id);
    if (!package) {
      return res.status(404).json({ message: 'Package not found' });
    }
    await package.destroy();
    res.json({ message: 'Package deleted' });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
};

File: 17_Code-Structure/controller/package_controller.go
package controllers

import (
	"net/http"
	"strconv"

	"code-structure/models"

	"github.com/jinzhu/gorm"
	"github.com/labstack/echo/v4"
)

type PackageController struct {
	DB *gorm.DB
}

func (pc *PackageController) GetAllPackages(c echo.Context) error {
	var packages []models.Package
	pc.DB.Find(&packages)
	return c.JSON(http.StatusOK, packages)
}

func (pc *PackageController) GetPackageByID(c echo.Context) error {
	id, _ := strconv.Atoi(c.Param("id"))
	var pkg models.Package
	if err := pc.DB.First(&pkg, id).Error; err != nil {
		return c.JSON(http.StatusNotFound, echo.Map{"message": "Package not found"})
	}
	return c.JSON(http.StatusOK, pkg)
}

func (pc *PackageController) CreatePackage(c echo.Context) error {
	var pkg models.Package
	if err := c.Bind(&pkg); err != nil {
		return c.JSON(http.StatusBadRequest, err)
	}
	pc.DB.Create(&pkg)
	return c.JSON(http.StatusCreated, pkg)
}

func (pc *PackageController) UpdatePackage(c echo.Context) error {
	id, _ := strconv.Atoi(c.Param("id"))
	var pkg models.Package
	if err := pc.DB.First(&pkg, id).Error; err != nil {
		return c.JSON(http.StatusNotFound, echo.Map{"message": "Package not found"})
	}

	if err := c.Bind(&pkg); err != nil {
		return c.JSON(http.StatusBadRequest, err)
	}
	pc.DB.Save(&pkg)
	return c.JSON(http.StatusOK, pkg)
}

func (pc *PackageController) DeletePackage(c echo.Context) error {
	id, _ := strconv.Atoi(c.Param("id"))
	if err := pc.DB.Delete(&models.Package{}, id).Error; err != nil {
		return c.JSON(http.StatusNotFound, echo.Map{"message": "Package not found"})
	}
	return c.NoContent(http.StatusNoContent)
}
